cms/djangoapps/contentstore/outlines_backfill.py

```diff
@@ -0,0 +1,138 @@
+    """
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.list_display_links = None
+        self.readonly_fields = [f.name for f in self.model._meta.get_fields()]
+
+    def get_actions(self, request):
+        actions = super().get_actions(request)
+        if 'delete_selected' in actions:
+            del actions["delete_selected"]
+        return actions
+
+    def has_add_permission(self, request):
+        return False
+
+    def has_delete_permission(self, request, obj=None):  # pylint: disable=unused-argument
+        return False
+
+    def save_model(self, request, obj, form, change):
+        pass
+
+    def delete_model(self, request, obj):
+        pass
+
+    def save_related(self, request, form, formsets, change):
+        pass
+
+
+class CourseOutlineBackfill(CourseOverview):
+    """
+    Proxy model for CourseOverview.
+
+    Does *not* create/update/delete CourseOverview objects - only reads the objects.
+    Uses the course IDs of the CourseOverview objects to determine which course
+    outlines to re-build/backfill.
+    """
+    class Meta:
+        proxy = True
+
+    def __str__(self):
+        """Represent ourselves with the course key."""
+        return str(self.id)
+
+    @classmethod
+    def get_course_outline_ids(cls):
+        """
+        Returns all the CourseOverview object ids.
+        """
+        return cls.objects.values_list('id', flat=True)
+
+
+def backfill_course_outlines_subset(modeladmin, request, queryset):
+    """
+    Create a celery task to backfill a single course outline for each passed-in course key.
+
+    If the number of passed-in course keys is above a threshold, then instead create a celery task which
+    will then create a celery task to backfill a single course outline for each passed-in course key.
+    """
+    all_course_keys_qs = queryset.values_list('id', flat=True)
+    if len(all_course_keys_qs) >= TASK_TO_LAUNCH_TASKS_THRESHOLD:
+        # Create the celery task-creating task.
+        backfills = len(all_course_keys_qs)
+        update_multiple_outlines_from_modulestore_task.delay([str(course_key) for course_key in all_course_keys_qs])
+        log.info("Queuing task to create outline creation tasks for all (%s) courses.", backfills)
+    else:
+        # Create a separate celery task for each course outline requested.
+        backfills = 0
+        for course_key in all_course_keys_qs:
+            if key_supports_outlines(course_key):
+                log.info("Queuing outline creation for %s", course_key)
+                update_outline_from_modulestore_task.delay(str(course_key))
+                backfills += 1
+            else:
+                log.info("Outlines not supported for %s - skipping", course_key)
+    plural_suffix = 's' if backfills > 1 else ''
+    modeladmin.message_user(
+        request,
+        f"Successfully requested {backfills} course outline{plural_suffix} to be backfilled."
```

Comment:
Nit: I've been pretty bad about i18n in Django admins, but if you want to do the right thing for this kind of pluralization, please see: https://edx.readthedocs.io/projects/edx-developer-guide/en/latest/internationalization/i18n.html#singular-vs-plural

Reply:
Yes - you're right. I'm totally not used to writing end-user facing messages anymore. I'll fix this.